### DISCUSSION TITLE
VCS provider setup

### DISCUSSION THREAD

**Discussion by erick-prosimo** [terrakube-io/terrakube#299] [Q&A] [answered]
I am trying to add a VCS provider, and when looking at the call being made through the UI I am getting a 403. I have validated I am an admin as I can authenticate, create workspaces, etc. Currently when I go to add a VCS provider it just does nothing once I click "connect and continue"

{errors: [{detail: "CreatePermission Denied"}]}
errors
: 
[{detail: "CreatePermission Denied"}]
0
: 
{detail: "CreatePermission Denied"}
detail
: 
"CreatePermission Denied"

**Comment by alfespa17** [comment, marked as answer]
The terrakube admin group can just create organizations and handle team access.

If you want to work inside any organization adding/updating workspaces, modules, vcs connections or templates you will have to add the "team" inside the organization setting and grant the require permissions. 

For example:
![image](https://user-images.githubusercontent.com/4461895/216147296-8b9a19ef-bace-4b3d-a600-2e7031453bb2.png)

![image](https://user-images.githubusercontent.com/4461895/216148664-fc12d9b0-bd8d-4a8f-88dc-73176b000709.png)

![image](https://user-images.githubusercontent.com/4461895/216148005-aee3edd7-0d15-442d-aef2-7934acd84989.png)

In your case you will have to add the "AD Group name" when you create the team.

**Reply by erick-prosimo** [reply]
Got it, makes sense. 1 thing is I am seeing this for the generated callback URL: 
https://terrakube-callback.<mydomain> and this URL does not exist.

**Reply by alfespa17** [reply]
Can you share a screenshot?

**Reply by erick-prosimo** [reply]
![image](https://user-images.githubusercontent.com/100948345/216156523-ea31c0c3-b26e-4d2f-9fed-558b910b7b34.png)

**Reply by erick-prosimo** [reply]
FYI, if I switch the callback url to be terrakube-api.<mydomain> I get this:  {"errors":[{"detail":"Found undefined keys in request: code"}]}

**Reply by alfespa17** [reply]
The callback URL should be the Terrakube API url.

You should be able to see the UI setup with this path.

Example UI path: 
```bash
https://3000-azbuilder-terrakube-ub9slgqr41v.ws-us85.gitpod.io/env-config.js
```

![image](https://user-images.githubusercontent.com/4461895/216158228-fb74eef7-1e91-4b6c-a052-0f49c39a700a.png)

Can you check the values please

**Reply by erick-prosimo** [reply]
Yup, all have the correct values.

**Reply by erick-prosimo** [reply]
Just tried again and still seeing the bad URL "terrakube-callback" being generated. That is not in my env-config. All my values are the correct terrakube-api/reg/ui

**Reply by alfespa17** [reply]
What is the version of the UI that you are using in the helm chart? Are you using the 2.10.0? 

```yaml
## UI Properties
ui:
  enabled: true
  version: "2.10.0"
  replicaCount: "1"
  serviceType: "ClusterIP"
```

Could you check this guide? https://docs.terrakube.org/user-guide/vcs-providers/github.com

**Reply by erick-prosimo** [reply]
Yes, 2.10.0 for everything. I follow that guide and that URL is generated. I have no control over the URL that is generated by Terrakube....

**Reply by alfespa17** [reply]
This is a bug, I found the issue in the UI code.

https://github.com/AzBuilder/terrakube/blob/main/ui/src/domain/Settings/VCS.jsx#L85

```javascript
const getCallBackUrl = (id) => {
    return `${window._env_.REACT_APP_TERRAKUBE_API_URL.replace("api","callback")}vcs/${id}`;
  }
```

I will have to update the logic of the callback URL logic, I never had this issue when I was testing in Gitpod because the API url never have the word "api"  in the hostname.

**Reply by alfespa17** [reply]
I have created a new version that fixes the issue with the callback URL just update to version 2.10.1 running a helm upgrade.
